src/vehicle_gui.cpp

```diff
@@ -1611,27 +1810,47 @@ struct VehicleListWindow : public BaseVehicleListWindow {
 	{
 		switch (widget) {
 			case WID_VL_SORT_ORDER: // Flip sorting method ascending/descending
-				this->vehicles.ToggleSortOrder();
+				this->vehgroups.ToggleSortOrder();
 				this->SetDirty();
 				break;
 
+			case WID_VL_GROUP_BY_PULLDOWN:// Select sorting criteria dropdown menu
+				ShowDropDownMenu(this, this->vehicle_group_by_names, this->grouping, WID_VL_GROUP_BY_PULLDOWN, 0, 0);
+				return;
+
 			case WID_VL_SORT_BY_PULLDOWN:// Select sorting criteria dropdown menu
-				ShowDropDownMenu(this, this->vehicle_sorter_names, this->vehicles.SortType(), WID_VL_SORT_BY_PULLDOWN, 0,
+				ShowDropDownMenu(this, this->GetVehicleSorterNames(), this->vehgroups.SortType(), WID_VL_SORT_BY_PULLDOWN, 0,
 						(this->vli.vtype == VEH_TRAIN || this->vli.vtype == VEH_ROAD) ? 0 : (1 << 10));
 				return;
 
 			case WID_VL_LIST: { // Matrix to show vehicles
 				uint id_v = this->vscroll->GetScrolledRowFromWidget(pt.y, this, WID_VL_LIST);
-				if (id_v >= this->vehicles.size()) return; // click out of list bound
-
-				const Vehicle *v = this->vehicles[id_v];
-				if (!VehicleClicked(v)) {
-					if (_ctrl_pressed) {
-						ShowCompanyGroupForVehicle(v);
-					} else {
-						ShowVehicleViewWindow(v);
+				if (id_v >= this->vehgroups.size()) return; // click out of list bound
+
+				const GUIVehicleGroup &vehgroup = this->vehgroups[id_v];
+				switch (this->grouping) {
+					case GB_NONE: {
+						const Vehicle *v = vehgroup.GetSingleVehicle();
+						if (!VehicleClicked(v)) {
+							if (_ctrl_pressed) {
+								ShowCompanyGroupForVehicle(v);
+							} else {
+								ShowVehicleViewWindow(v);
+							}
+						}
+						break;
 					}
+
+					case GB_SHARED_ORDERS:
+						assert(vehgroup.NumVehicles() > 0);
+						// TODO: support ctrl-click for cloning shared orders
+						// this might be hard, because there is currently no way to clone shared orders except by supplying a vehicle (but we don't want to enable other kinds of vehicle functionality)
```

Comment:
I don't really like `TODO`s in code. It is an idea you currently have, which might faint over time. And it becomes more and more difficult over time to understand what people meant. So either we plan to add this, and do that, or we don't. But mentioning possible feature additions .. I suggest we just remove the comment ;)

Reply:
You did not pick up this comment (and the similar other two); you still working on that, or you don't agree? :)

Reply:
Still working on that, didn't have the time to do it this morning :)

Reply:
No worries, didn't want to rush you :) Just sometimes it is hard to see what is and isn't done :) Tnx!

Reply:
Done now :)